torch/_dynamo/optimizations/training.py

```diff
@@ -260,42 +260,139 @@ def __call__(self, gm: torch.fx.GraphModule, example_inputs):
-    # execute function.
+def _has_incompatible_cudagraph_ops(gm):
+    from torch._inductor.utils import (
+        has_incompatible_cudagraph_ops as has_incompatible_cudagraph_ops_inductor,
+    )
+
+    result = has_incompatible_cudagraph_ops_inductor(gm)
+    if result:
+        return result
+
+    incompatible = [
+        "aten.index_put.default",
+    ]
+    for node in gm.graph.nodes:
+        if str(node.target) in incompatible:
+            return True
+
+    # A workaround for
+    # https://github.com/csarofeen/pytorch/issues/2106
+    is_any_ones = any(
+        node.target == torch.ops.aten.ones.default for node in gm.graph.nodes
+    )
+    is_any_zeros = any(
+        node.target == torch.ops.aten.zeros.default for node in gm.graph.nodes
+    )
+    if is_any_ones and is_any_zeros:
+        return True
+
+    return False
+
+
+def _replace_cpu_with_cuda(gm):
+    for node in gm.graph.nodes:
+        if node.op == "call_function":
+            new_kwargs = dict(node.kwargs)
+            if new_kwargs.get("device", False) and new_kwargs["device"].type == "cpu":
+                new_kwargs["device"] = torch.device("cuda")
+            node.kwargs = new_kwargs
+    gm.recompile()
+    return gm
+
+
+def prims_executor(gm, inputs, *, executor, num_fixed=0, cudagraphs=False):
+    """This function is called once per forward/backward pass of a graph in AOT
+    Autograd. We use it to set up the nvFuser-specific FX graph and return
+    execute function."""
+    from functorch.compile import make_boxed_func
+
+    from torch._inductor.compile_fx import align_inputs, cudagraphify
     from torch._prims.context import TorchRefsNvfuserCapabilityMode
     from torch._prims.executor import execute
     from torch.fx.experimental.proxy_tensor import make_fx
 
+    # cudagraphify fails if intermediate tensors are CPU and there's an attempt
+    # to send them to CUDA
+    if cudagraphs:
+        gm = _replace_cpu_with_cuda(gm)
+
     # First we trace the graph conditionally decomposing nodes
     # that can be sent to the nvfuser executor
     with TorchRefsNvfuserCapabilityMode():
         prim_gm = make_fx(gm)(*inputs)
 
     # Then we return a callable that executes the "prim_gm" graph
-    return partial(execute, prim_gm, executor=executor)
+    run = make_boxed_func(partial(execute, prim_gm, executor=executor))
 
+    if _has_incompatible_cudagraph_ops(prim_gm) or not cudagraphs:
+        return run
 
-def create_nvprims_backend(*, executor):
+    try:
+        # Inductor's cudagraphify has hardcoded alignment constant for Triton
+        # We don't need it here, so we just set it to 1
+        old_alignment = torch._inductor.compile_fx.ALIGNMENT
+        torch._inductor.compile_fx.ALIGNMENT = 1
+        cudagraphed_fn = cudagraphify(run, inputs, range(num_fixed))
+        result = align_inputs(cudagraphed_fn, inputs, range(num_fixed))
+        result._boxed_call = True
+    finally:
```

Comment:
nitpick, should we add a catch here where we would return `run` when cudagraphify fails?

Reply:
It's difficult to go back and actually fix problems once we add a catch-all fallback. I would prefer to have explicit checks for failing conditions we discover.

Reply:
> It's difficult to go back and actually fix problems once we add a catch-all fallback.

Yeah, we have the same problem with speculative lowering~~
I'd argue that's a better user experience :wink: We can env-var guard that fallback behavior for debugging.

We don't have to do it here right now, as you mentioned `cudagraphify doesn't fail immediately here`. Let's revisit this when we run into UX issue then.